k2/csrc/connect.cu

```diff
@@ -0,0 +1,519 @@
+/**
+ * Copyright      2021  Xiaomi Corporation (authors: Wei Kang)
+ *
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <limits>
+#include <vector>
+
+#include "k2/csrc/array_ops.h"
+#include "k2/csrc/context.h"
+#include "k2/csrc/device_guard.h"
+#include "k2/csrc/fsa_algo.h"
+#include "k2/csrc/fsa_utils.h"
+#include "k2/csrc/thread_pool.h"
+
+namespace k2 {
+
+class Connecter {
+ public:
+  /**
+     Connector object.  You should call Connect() after
+     constructing it.  Please see Connect() declaration in header for
+     high-level overview of the algorithm.
+
+       @param [in] fsas    A vector of FSAs; must have 3 axes.
+   */
+  explicit Connecter(FsaVec &fsas) : c_(fsas.Context()), fsas_(fsas) {
+    K2_CHECK_EQ(fsas_.NumAxes(), 3);
+    int32_t num_states = fsas_.shape.TotSize(1);
+    accessible_ = Array1<char>(c_, num_states, 0);
+    coaccessible_ = Array1<char>(c_, num_states, 0);
+  }
+
+  /*
+    Computes the next batch of states
+         @param [in] cur_states  Ragged array with 2 axes, containing
+    state-indexes (idx01) into fsas_.
```

Comment:
Usually, it's better if we can add comments with the index convention here to describe the shape of `cur_states`, such as `[fsa][state]`